cni/pkg/install/install.go

```diff
@@ -176,6 +170,64 @@ func (in *Installer) Cleanup() error {
 	return nil
 }
 
+// tryRemoveChainedConfig reads the current shared CNI plugin config,
+// removes the Istio plugin entry, then writes the file back out.
+//
+// If the file changes between the time when we read it and attempt to write it back out,
+// then we do not write and return an error instead - retries should be performed in this case.
+//
+// This is required because in chained mode, multiple node agents with their own plugins may be
+// watching/writing/contending over the same file we are.
+func (in *Installer) tryRemoveChainedConfig() error {
+	installLog.Infof("Removing Istio CNI config from CNI config file: %s", in.cniConfigFilepath)
+	// Corner case handling - we read the file into memory, excise our config, and overwrite the result
+	// back out to the file. 99% of the time this is fine - but if we are chained with another node agent
+	// doing something similar, the other node agent can read/excise/write the file in between the
+	// time when we read it in and write it back out - in that case we end up writing invalid node config.
+	//
+	// So, do a simple check to make sure the file hasn't changed since we initially read it. If it has, discard and retry.
+	cniConfigInitial, err := os.Stat(in.cniConfigFilepath)
+	if err != nil {
+		return err
+	}
+	// Read JSON from CNI config file
+	cniConfigMap, err := util.ReadCNIConfigMap(in.cniConfigFilepath)
+	if err != nil {
+		return err
+	}
+	// Find Istio CNI and remove from plugin list
+	plugins, err := util.GetPlugins(cniConfigMap)
+	if err != nil {
+		return fmt.Errorf("%s: %w", in.cniConfigFilepath, err)
+	}
+	for i, rawPlugin := range plugins {
+		plugin, err := util.GetPlugin(rawPlugin)
+		if err != nil {
+			return fmt.Errorf("%s: %w", in.cniConfigFilepath, err)
+		}
+		if plugin["type"] == "istio-cni" {
+			cniConfigMap["plugins"] = append(plugins[:i], plugins[i+1:]...)
+			break
+		}
+	}
+
+	cniConfig, err := util.MarshalCNIConfig(cniConfigMap)
+	if err != nil {
+		return err
+	}
+
+	if stat, err := os.Stat(in.cniConfigFilepath); err != nil ||
```

Comment:
Does this really meaningfully change anything? We are still TOCTOU-ing (as you noted). It seems like the time for the IO is many orders of magnitude larger than the time of work (which is trivial, near instant). So I don't see this check adding value beyond complexity.

Reply:
We've seen it happen with multiple plugins/agents, where the istio installer copies an out-of-date config back out to the node, with spurious entries in it, causing subsequent pods to not schedule.

It's admittedly a corner case. This tightens the window. Our options are 

1. Leave it.
2. Make it better.
3. Change the spec and obviate the problem.

More/better ideas welcome, I guess.

Reply:
I like tightening the window, I just don't think this actually does it.

With some made-up timings:

Before:
1. 100ms: Read config
2. 1ms: mutate config
3. 100ms: write config

There is a 101ms window of conflict. Any write in this window we will revert.

Now:

1. 50ms: stat config
2. 100ms: read config
3. 1ms: mutage config
4. 50ms: stat config -> maybe exit
5. 100ms: write config

There is a 101ms window of conflict which we will retry on. There is a 150ms window on conflict (start of 2nd stat -> end of write) during which we will revert and not retry.

Seems worse actually? A bit vague on when exactly during the IO actions changes to the underlying files apply, though, so that isn't exactly correct probably

Reply:
It's probably more of a wash given the actual IO timings. This is also mostly an issue when multiple things are performing uninstall logic/cleanup at the same time. But correct. Once we hand over to `AtomicWrite` it's out of our hands. This is more about making sure `istio-cni` isn't at fault.

Other option is make a backup, do the ops, then reread _after_ write to make sure the file we have on disk is what we expect. That gets messy though.

Reply:
If you aren't fond of this approach and we don't have better options (I honestly can't think of many), I'm down with dropping this PR and basically flagging the underlying issue as WONTFIX.

Reply:
I'm raising https://github.com/istio/istio/issues/46848 so the underlying issue isn't lost, and closing this PR (since by itself it's not an effective fix)